#nnVar v.1.0
Contact: zhangyc@mail.bnu.edu.cn
Date: 2016-05-03# About annVar

annVar is a tool to utilize update-to-date information to functionally annotate genetic variants detected from diverse genomes (e.g. plants, Escherichia coli, yeast and many others).

## Learning annVar

Currently, annVar should perform based on Gbowse database:

- First, load the reference genome annotation into a Gbrowse database.

- Second, load lists of variants (including chromosome, start position, end position, reference nucleotide and observed nucleotides) into the Gbrowse database.

-  Third, run annVar to identify whether SNPs or indels cause protein coding changes and the amino acids that are affected.


## System Requirements

annVar has been tested on systems running Linux on an x86_64 architecture.
Compiling the program requires GCC v4.4.6 or greater.


## Installation

Run 'make' in the source directory


## Version

Version: v.1.0
Contact: zhangyc@mail.bnu.edu.cn
Date: 2016-05-03
